README.md

```diff
@@ -1 +1,130 @@
-# earthstar-fetch
\ No newline at end of file
+# earthstar-fetch
+
+A `fetch()` interface for talking to EarthStar
+
+## Goals
+
+- Stick to existing conventions in HTTP-land (body, methods, prefer standard headers)
+- GET data from earthstar workspaces
+- GET an author address for a given name
+- PUT data into a workspace
+- Auto-generate identity keys from a 4 character name
+- Search via querystring params and path prefixes
+
+## Dream API
+
+Basic URL format:
+
+```
+earthstar://+gardening.xxxxxxxxx/path/to/key.json
+```
+
+GET document
+
+```JavaScript
+const res = await fetch('earthstar://+gardening.xxxxx/example')
+
+// Get the timestamp for when it was last updated
+res.headers.get('mtime') === "1231231231231"
+// Get the author for this document
+res.headers.get('earthstar-author') === '@bbbb.xxxxxxxx'
+// Get the signature for this document, in case you want it 🤷
+res.headers.get('earthstar-signature') === 'xxxxxxx'
```

Comment:
So all the document metadata goes into the headers?

Here's a [full list of metadata fields](https://github.com/earthstar-project/earthstar/blob/master/src/util/types.ts#L154-L164)

Should at least also include `deleteAfter` so you know if the document is ephemeral or not.

Reply:
Is this something we could do in the content type accept header or in the query string?

Maybe `/path/to/thing?document` will return the full document's JSON kinda like how doing `?document` in a search will return the full documents?

Reply:
### Asking for metadata vs asking for content

Sometimes we want the "metadata", sometimes the "content", sometimes both ("document").  Sometimes we want the metadata in its original JSON format for ease of use, and maybe sometimes it can be in HTTP headers instead.

The various ways we could do this...

1. `HEAD` to get JSON metadata without content, `GET` to get... content in the body and metadata in the headers?
2. Use request headers `Accept: earthstar/content`, `Accept: earthstar/metadata`, etc
3. `/path/to/thing?XXX` where `XXX` is one of `content`, `metadata`, `document`, defaulting to content if missing
4. `/path/to/thing?type=XXX`, similar above except this can be combined with the other query options.

I like the last one, it seems a natural fit to combine it with the rest of the query options.  It's also nice to keep it all in the URL so you don't have to mess around with request headers.

So the combined format would be:

`earthstar://  WORKSPACE  /  PATH  ?  QUERY`

...where PATH is either a specific actual path, or empty (to query the whole workspace), or maybe something like `/foo*` or `/foo/` as a shorthand for a path prefix query.

And QUERY will now have an extra field, `type`, which can be `path | metadata | content | document`, and it should also have a `sort` option which will be something like `path | newest | oldest`.

Reply:
...and if the query gets multiple results, how do we combine them:

* `type=path` - just plain strings separated by `\n`?  or a JSON array?
* `type=content` - concatenate the raw content, separated by `\n`
* `type=metadata` and `type=document` - these are JSON objects, so return a JSON array of them for easy parsing

Reply:
I really like the `?type=` option from that list. I think one advantage over special methods or accept headers is that we can link to it more easily.

I think with path, having it as JSON would be most useful, unless there's an accept for `text/html` so that it can render a basic index page. Newline separated is good too, but I think it'd make life a lot easier on app devs to be able to use `res.json()`.

Regarding content, I could imagine there being continuous chunks or binary data where you wouldn't want an extra `\n`. 🤔

The sort stuff sounds gr8.

Reply:
> one advantage over special methods or accept headers is that we can link to it more easily.

Yeah, and also more accessible to users via the URL field of the browser.

> binary data where you wouldn't want an extra `\n` 🤔

You could specify the array or concatenation style too.  `type=path&concatWith=newline` or `concatWith=raw` or `concatWith=jsonArray` ? 🤷 ... rendered HTML index page also almost belongs in there but only for paths, `concatWith=anchorTags`, 🤷🤷

If you were concatenating html you might want to wrap or separate each piece with a certain tag... `wrapWith=p`... `concatWith=hr`... this is getting too fancy :)

Reply:
Hmmmm, it does sound kinda fancy... 😅 How do you feel about the initial implementation being straight concatenation without delimiters or wrappers and figuring out the fancier stuff later?

Reply:
Ok with me!  For some things like listing paths we definitely need a delimieter, `\n` seems the obvious choice, otherwise maybe just try something and we'll run from there.